sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/NuGetMetadataReference.cs

```diff
@@ -69,15 +69,28 @@ public static MetadataReference[] Create(string packageId, string packageVersion
         {
             EnsurePackageIsInstalled(packageId, packageVersion);
 
+            var allowedNugetLibDirectoriesByPreference = allowedNugetLibDirectories.
+                Zip(Enumerable.Range(0, allowedNugetLibDirectories.Count), (folder, priority) => new { folder, priority });
             var packageDirectory = GetNuGetPackageDirectory(packageId, packageVersion);
-            var matchingDlls = Directory.GetFiles(packageDirectory, "*.dll", SearchOption.AllDirectories)
+            var matchingDllsGroups = Directory.GetFiles(packageDirectory, "*.dll", SearchOption.AllDirectories)
                 .Select(path => new FileInfo(path))
-                .GroupBy(file => file.Directory.Name)
-                .Where(group => allowedNugetLibDirectories.Any(y => group.Key.StartsWith(y)) || group.Key.EndsWith(".dll"))
-                .OrderByDescending(group => group.Key)
-                .First();
+                .GroupBy(file => file.Directory.Name).ToArray();
+            IGrouping<string, FileInfo> selectedGroup;
+            if (matchingDllsGroups.Length == 1 && matchingDllsGroups[0].Key.EndsWith(".dll"))
+            {
+                selectedGroup = matchingDllsGroups[0];
+            }
+            else
+            {
+                selectedGroup = matchingDllsGroups.Join(allowedNugetLibDirectoriesByPreference,
```

Comment:
This LINQ statement is pretty complex and I think it deserves a separate method with appropriate name